README.md
README.txt for oversee
----------------------

"oversee" is an embeddable unit-test package inspired by "[seatest](https://github.com/keithn/seatest)".


CHANGES.txt
Changes in oversee
------------------

V0.91a
_x_	Module created


tests/001-main/001-main.c
/* 001-main.c	Sun Mar  3 2013 18:06:54 tmm */

/*

Module:  001-main.c

Function:
	Test for oversee -- simplest possible configuration.

Version:
	V0.91a	Sun Mar  3 2013 18:06:54 tmm	Edit level 1

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY  14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore.
 
Author:
	Terry Moore	March 2013

Revision history:
   0.91a  Sun Mar  3 2013 18:06:54  tmm
	Module created.

*/

#include <stdio.h>
#include <stdlib.h>

#include "oversee.h"

/****************************************************************************\
|
|		Manifest constants & typedefs.
|
|	This is strictly for private types and constants which will not 
|	be exported.
|
\****************************************************************************/

static OVERSEE_SUITE_DRIVER_FN SuiteDriver;

/****************************************************************************\
|
|	Read-only data.
|
|	If program is to be ROM-able, these must all be tagged read-only 
|	using the ROM storage class; they may be global.
|
\****************************************************************************/



/****************************************************************************\
|
|	VARIABLES:
|
|	If program is to be ROM-able, these must be initialized
|	using the BSS keyword.  (This allows for compilers that require
|	every variable to have an initializer.)  Note that only those 
|	variables owned by this module should be declared here, using the BSS
|	keyword; this allows for linkers that dislike multiple declarations
|	of objects.
|
\****************************************************************************/

int main(
	int argc,
	char **argv
	)
	{
	OVERSEE_SUITE *pSuite;
	OVERSEE_BOOL fResult;
	
	pSuite = Oversee_StaticInit();

	fResult = OverseeSuite_RunTests(pSuite, SuiteDriver, NULL);

	return fResult ? EXIT_SUCCESS : EXIT_FAILURE;
	}

void
SuiteDriver(
	OVERSEE_SUITE *pSuite,
	void *pContext_void
	)
	{
	/* nothing */
	}


tests/000-include/oversee.h.c
#include "../../i/oversee.h"


tests/000-include/oversee_env.h.c
#include "../../i/oversee_env.h"


i/oversee_env.h
/* oversee_env.h	Sun Mar  3 2013 18:35:40 tmm */

/*

Module:  oversee_env.h

Function:
	Compiler and environment adaptation for the oversee unit-test package.

Version:
	V0.91a	Sun Mar  3 2013 18:35:40 tmm	Edit level 2

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY 14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore
 
Author:
	Terry Moore	February 2013

Revision history:
   0.91a  Sat Feb 16 2013 05:54:15  tmm
	Module created.

*/

#ifndef OVERSEE_ENV_H__		/* prevent multiple includes */
#define OVERSEE_ENV_H__

/****************************************************************************\
|
|	Compiler and environment adaptation
|
| Header file dependencies: <stddef.h>, ISO C99 compliant.  This means we're
| allowed to use size_t, ptrdiff_t, wchar_t, NULL and offsetof().
|
| Oversee uses the term "ABI keyword" to denote the special syntax used
| by many C compilers to tag functions and function prototypes as using
| a particular machine-language calling sequence.  (ABI stands for
| "application binary interface".)  One of the purposes of this module is
| to define appropriate OVERSEE_... symbols that expand to the appropriate
| compiler-specific tag for functions of given types.
|
| Compile-time parameters:
|
|	_MSC_VER		if defined, triggers MSFT C awareness
|
|	OVERSEE_ABI_STD		if defined, is the ABI keyword for
|				decorating the fixed-signature ABIs.
|
|	OVERSEE_ABI_VARARGS	if defined, is the ABI keyword for
|				decorating the variadic ABIs.
|
|	OVERSEE_RESTRICT	if defined, is the ABI keyword that
|				provides the C99 "restrict" keyword
|				functionality.
|
|	OVERSEE_STDDEF_H	if defined, this is the name of
|				a header file to use instead of
|				<stddef.h>.  It is normally defined
|				to the compiler using
|				-D OVERSEE_STDDEF_H=local_stddef.h --
|				without quotes.  This syntax is
|				easiest to use, but doesn't allow
|				for angle-brackets. It will be
|				#included using "name" notation.
|
|	OVERSEE_STDDEF_H_META	if defined, this is the name of
|				a header file to use instead of <stddef.h>
|				It will be #included using <metabracket>
|				notation.
|
| Post conditions of this module:
|	<stddef.h> has been included.
|
|	OVERSEE_ABI_STD is always defined, and should be used when declaring
|	function types for functions that take a fixed number of parameters.
|	It should be used as follows:
|
|	typedef {type} OVERSEE_ABI_STD {FNTYPENAME}( {params} );
|
|	OVERSEE_ABI_VARARGS is always defined, and should be used when
|	declaring function types for functions that take a variable number of
|	parameters.  It should be used as follows:
|
|	typedef {type} OVERSEE_ABI_VARARGS {FNTYPENAME} ( {params}, ... );
|
|	This syntax is annoying, but is required in order to allow oversee
|	code to be safely embedded in the widest variety of environments.
|
|	OVERSEE_STRING() is always defined as a macro that returns the
|	stringification of its argument, without macro expansion.
|
|	OVERSEE_STRINGVAL() is always defined as a macro that returns the
|	stringification of its argument, after macro expansion.
|
|	OVERSEE_CONCAT3(x,y,z) is defined as a macro that returns the
|	paste of its arguments, without macro exapansion.
|
|	OVERSEE_CONCATVAL3(x,y,z) is defined as a macro that returns the
|	paste of its arguments, after macro expansion.
|
\****************************************************************************/

/****************************************************************************\
|
|	Define the ABI symbols
|
\****************************************************************************/

#ifndef OVERSEE_ABI_STD
# ifdef _MSC_VER
#  define OVERSEE_ABI_STD	__stdcall
# else
#  define OVERSEE_ABI_STD	/* nothing */
# endif
#endif

#ifndef OVERSEE_ABI_VARARGS
# ifdef _MSC_VER
#  define OVERSEE_ABI_VARARGS	__cdecl
# else
#  define OVERSEE_ABI_VARARGS	/* nothing */
# endif
#endif

/****************************************************************************\
|
|	Define the "restrict" keyword
|
\****************************************************************************/

/*
|| Many C compilers implement the restrict keyword, but not all. Rather
|| than using "restrict" and hoping for the best, we simply define an
|| oversee variant that can be overridden if needed.
*/
#ifndef OVERSEE_RESTRICT
# if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
#   define OVERSEE_RESTRICT	restrict
# elif defined __GNUC__
#   define OVERSEE_RESTRICT	__restrict__
# else
#   define OVERSEE_RESTRICT	/* nothing */
# endif
#endif

/****************************************************************************\
|
|	Define the string-manipulation macros
|
\****************************************************************************/

#define	OVERSEE_STRING(x)	#x
#define	OVERSEE_STRINGVAL(x)	OVERSEE_STRING(x)
#define	OVERSEE_CONCAT3(x,y,z)	x##y##z
#define	OVERSEE_CONCATVAL3(x,y,z) OVERSEE_CONCAT3(x,y,z)

/****************************************************************************\
|
|	Include <stddef.h> or equivalent.
|
| Post-condition: types ptrdiff_t, size_t, wchar_t are defined
|		  macros NULL, offsetof(t, member) are defined
|
\****************************************************************************/

#if defined(OVERSEE_STDDEF_H_META)
# if defined(OVERSEE_STDDEF_H)
#  error "Do not define both OVERSEE_STDDEF_H and OVERSEE_STDDEF_H_META"
# endif
# include OVERSEE_CONCATVAL3(<,OVERSEE_STDDEF_H_META,>)
#elif defined(OVERSEE_STDDEF_H)
# include OVERSEE_STRINGVAL(OVERSEE_STDDEF_H)
#else
# include <stddef.h>
#endif

/****************************************************************************\
|
|	Define the begin/end declaration tags for C++ co-existance
|
\****************************************************************************/

#ifdef __cplusplus
# define OVERSEE_BEGIN_DECLS	extern "C" {
# define OVERSEE_END_DECLS	}
#else
# define OVERSEE_BEGIN_DECLS	/* nothing */
# define OVERSEE_END_DECLS	/* nothing */
#endif

/****************************************************************************\
|
|	Get set up for SAL-style annotations
|
| Postcondition:
|	OVERSEE_PARAM_USE_SAL is defined and numeric
|
|	OVERSEE_SAL_Function_class(n) is defined and like _Function_class_(n)
|	OVERSEE_SAL_In, OVERSEE_SAL_Out, OVERSEE_SAL_In_z,
	OVERSEE_SAL_Out_opt, OVERSEE_SAL_Out_writes_z(nChar)
|
\****************************************************************************/

#ifndef OVERSEE_PARAM_USE_SAL
# if defined(_MSCVER)
#  define OVERSEE_PARAM_USE_SAL 1
# else
#  define OVERSEE_PARAM_USE_SAL 0
# endif
#endif

#if OVERSEE_PARAM_USE_SAL
# include <sal.h>

/* annotation defining a function class for role-checking */
# define OVERSEE_SAL_Function_class(n)	_Function_class_(n)
# define OVERSEE_SAL_In			_In_
# define OVERSEE_SAL_In_opt		_In_opt_
# define OVERSEE_SAL_In_opt_z		_In_opt_z_
# define OVERSEE_SAL_In_reads_or_z(c)	_In_reads_or_z_(c)
# define OVERSEE_SAL_In_z		_In_z_
# define OVERSEE_SAL_Out		_Out_
# define OVERSEE_SAL_Out_opt		_Out_opt_
# define OVERSEE_SAL_Out_writes_bytes(n) _Out_writes_bytes_(n)
# define OVERSEE_SAL_Out_writes_z(nc)	_Out_writes_z_(nc)
# define OVERSEE_SAL_Param(i)		_Param_(i)
# define OVERSEE_SAL_Post_invalid	_Post_invalid_
# define OVERSEE_SAL_Pre_defensive	_Pre_defensive_
# define OVERSEE_SAL_Printf_format_string _Printf_format_string_
# define OVERSEE_SAL_Ret_notnull	_Ret_notnull_
# define OVERSEE_SAL_Success(e)		_Success_(e)
#else /* !OVERSEE_PARAM_USE_SAL */

# define OVERSEE_SAL_Function_class(n)	/* nothing */
# define OVERSEE_SAL_In			/* nothing */
# define OVERSEE_SAL_In_opt		/* nothing */
# define OVERSEE_SAL_In_opt_z		/* nothing */
# define OVERSEE_SAL_In_reads_or_z(c)	/* nothing */
# define OVERSEE_SAL_In_z		/* nothing */
# define OVERSEE_SAL_Out		/* nothing */
# define OVERSEE_SAL_Out_opt		/* nothing */
# define OVERSEE_SAL_Out_writes_bytes(n) /* nothing */
# define OVERSEE_SAL_Out_writes_z(nc)	/* nothing */
# define OVERSEE_SAL_Post_invalid	/* nothing */
# define OVERSEE_SAL_Pre_defensive	/* nothing */
# define OVERSEE_SAL_Printf_format_string /* nothing */
# define OVERSEE_SAL_Ret_notnull	/* nothing */
# define OVERSEE_SAL_Success(e)		/* nothing */

#endif /* !OVERSEE_PARAM_USE_SAL */

/**** end of oversee_env.h ****/
#endif /* OVERSEE_ENV_H__ */


i/oversee_types.h
/* oversee_types.h	Sat Feb 16 2013 05:47:30 tmm */

/*

Module:  oversee_types.h

Function:
	Forward types for OVERSEE.

Version:
	V0.91a	Sat Feb 16 2013 05:47:30 tmm	Edit level 1

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY  14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore.
 
Author:
	Terry Moore	February 2013

Revision history:
   0.91a  Sat Feb 16 2013 05:47:30  tmm
	Module created.

*/

#ifndef OVERSEE_TYPES_H__		/* prevent multiple includes */
#define OVERSEE_TYPES_H__

#ifndef OVERSEE_ENV_H__
# include "oversee_env.h"
#endif

/****************************************************************************\
|
|	Forward types
|
| All tags end in {type}_s for struct, {type}_u for union, {type_e} for enum.
|
| We do not declare P{type} variants.
|
\****************************************************************************/

typedef	struct OVERSEE_CONTEXT_s	OVERSEE_CONTEXT;
typedef struct OVERSEE_SUITE_s		OVERSEE_SUITE;
typedef	struct OVERSEE_FIXTURE_s	OVERSEE_FIXTURE;
typedef	struct OVERSEE_TEST_s		OVERSEE_TEST;

typedef struct OVERSEE_PLATFORM_API_s	OVERSEE_PLATFORM_API;
typedef struct OVERSEE_RESULT_STREAM_API_s OVERSEE_RESULT_STREAM_API;

typedef	int OVERSEE_BOOL;

/*

Type:	OVERSEE_TEST_DURATION

Index:	Constant:	OVERSEE_TEST_DURATION_NIL
	String:		OVERSEE_TEST_DURATION_PRINT

Function:
	Represents a test duration, with designated "nil".

Description:
	OVERSEE_TEST_DURATION is a signed integer type that is used for
	representing a test duration.  Since we're designing for embedded
	systems, we don't actually specify how this maps to seconds or
	fractions thereof.  It's just got to be a signed integer value
	thats large enough to hold the measurements of interest.

	OVERSEE_TEST_DURATION_NIL is a distinguished value that can be
	used to indicate that a duration value is not specified.

	OVERSEE_TEST_DURATION_PRINT is a format specifier string (such as
	"ld") that will print a value other than NIL as a decimal number.

*/

#if defined(OVERSEE_TEST_DURATION_TYPE)
 typedef OVERSEE_TEST_DURATION_TYPE OVERSEE_OPTIONAL_TEST_DURATION;
# if ! (defined(OVERSEE_TEST_DURATION_NIL) && \
        defined(OVERSEE_TEST_DURATION_PRINT))
#   error "If you define OVERSEE_TEST_DURATION_TYPE, need _NIL and _PRINT too"
# endif
#else
  typedef long OVERSEE_OPTIONAL_TEST_DURATION;
# define OVERSEE_OPTIONAL_TEST_DURATION_NIL (-1l)
# define OVERSEE_OPTIONAL_TEST_DURATION_PRINT "ld"
#endif

/****************************************************************************\
|
|	The API function typedefs (used by clients).
|
| All function type names end with _FN. All callback function types end with
| _CB_FN. (Callback function types are functions that are passed as 
| parameters to APIs). 
|
| All API function type names are of the form OVERSEE_{object}_{method}_FN.
|
| The referenced object is the first parameter.
|
| All are conceptually designed as methods.
|
\****************************************************************************/

typedef 
OVERSEE_SAL_Function_class(OVERSEE_SUITE_SETUP_FN)
void OVERSEE_ABI_STD 
OVERSEE_SUITE_SETUP_FN(
	OVERSEE_SAL_In
		OVERSEE_SUITE *,
	OVERSEE_SAL_In_opt
		void *
	);

typedef	
OVERSEE_SAL_Function_class(OVERSEE_SUITE_TEARDOWN_FN)
void OVERSEE_ABI_STD 
OVERSEE_SUITE_TEARDOWN_FN(
	OVERSEE_SAL_In
		OVERSEE_SUITE *,
	OVERSEE_SAL_In_opt
		void *
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_FIXTURE_SETUP_FN)
void OVERSEE_ABI_STD 
OVERSEE_FIXTURE_SETUP_FN(
	OVERSEE_SAL_In
		OVERSEE_FIXTURE *,
	OVERSEE_SAL_In_opt
		void *
	);

typedef	
OVERSEE_SAL_Function_class(OVERSEE_FIXTURE_TEARDOWN_FN)
void OVERSEE_ABI_STD 
OVERSEE_FIXTURE_TEARDOWN_FN(
	OVERSEE_SAL_In
		OVERSEE_FIXTURE *,
	OVERSEE_SAL_In_opt
		void *
	);

typedef
OVERSEE_SAL_Function_class(OVERSEE_SUITE_DRIVER_FN)
void OVERSEE_ABI_STD
OVERSEE_SUITE_DRIVER_FN(
	OVERSEE_SAL_In
		OVERSEE_SUITE *		/* pSuite */,
	OVERSEE_SAL_In_opt
		void *			/* pContext */
	);

/**** end of oversee_types.h ****/
#endif /* OVERSEE_TYPES_H__ */


i/oversee_platform.h
/* oversee_platform.h	Thu Feb 28 2013 16:26:34 tmm */

/*

Module:  oversee_platform.h

Function:
	The interface to the OVERSEE platform

Version:
	V0.91a	Thu Feb 28 2013 16:26:34 tmm	Edit level 1

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY  14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore.
 
Author:
	Terry Moore	February 2013

Revision history:
   0.91a  Thu Feb 28 2013 16:26:34  tmm
	Module created.

*/

#ifndef OVERSEE_PLATFORM_H__		/* prevent multiple includes */
#define OVERSEE_PLATFORM_H__

#ifndef OVERSEE_PLATFORM_TYPES_H__
# include "oversee_platform_types.h"
#endif


struct OVERSEE_PLATFORM_API_s
	{
	size_t					sizeApis;
	OVERSEE_HPLATFORM			hPlatform;
	OVERSEE_PLATFORM_REFERENCE_FN		*pReference;
	OVERSEE_PLATFORM_DEREFERENCE_FN		*pDereference;
	OVERSEE_PLATFORM_SNPRINTF_FN		*pSnprintf;
	OVERSEE_PLATFORM_STRCMP_FN		*pStrcmp;
	OVERSEE_PLATFORM_STRLEN_FN		*pStrlen;
	OVERSEE_PLATFORM_STRNCMP_FN		*pStrncmp;
	OVERSEE_PLATFORM_STRSTR_FN		*pStrstr;
	OVERSEE_PLATFORM_PROTECTED_CALL_FN	*pProtectedCall;
	OVERSEE_PLATFORM_OPEN_RESULT_STREAM_FN	*pOpenResultStream;
	};


/**** end of oversee_platform.h ****/
#endif /* _OVERSEE_PLATFORM_H__ */


src/oversee_internals.h
/* oversee_internals.h	Sun Mar  3 2013 18:19:19 tmm */

/*

Module:  oversee_internals.h

Function:
	The internal structures for the Oversee unit test system.

Version:
	V0.91a	Sun Mar  3 2013 18:19:19 tmm	Edit level 1

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY  14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore.
 
Author:
	Terry Moore, Terrill Moore	March 2013

Revision history:
   0.91a  Sun Mar  3 2013 18:19:19  tmm
	Module created.

*/

#ifndef OVERSEE_INTERNALS_H__		/* prevent multiple includes */
#define OVERSEE_INTERNALS_H__

#ifndef OVERSEE_TYPES_H__
# include "oversee_types.h"
#endif

/****************************************************************************\
|
|	The overall OVERSEE_CONTEXT structure
|
\****************************************************************************/

struct OVERSEE_CONTEXT_s
	{
	/*
	|| There is intended to only one test suite per instance, but
	|| that's not actaully a requirement.  But since only one suite
	|| can be active per oversee instance, we only have memory for one.
	*/
	OVERSEE_SUITE	CurrentSuite;

	/*
	|| We expect to have multiple fixtures per instance, but only one
	|| active at a time. So we only have memory for one live instance.
	*/
	OVERSEE_FIXTURE	CurrentFixture;

	/*
	|| The platform pointers live separately so we don't have to have
	|| the structure in scope unless we really need it.
	*/
	OVERSEE_PLATFORM_API *pPlatform;

	/*
	|| The result stream API pointers similarly live separately, so we
	|| don't have to have things in scope.
	*/
	OVERSEE_RESULT_STREAM_API *pResultStream;

	/*
	|| we count the number of failures.
	*/
	unsigned	nTestsFailed;

	/*
	|| we count the number of tests run here.
	*/
	unsigned	nTestsRun;

	/*
	|| similarly, we count the number of fixtures failed/run
	*/
	unsigned	nFixturesFailed;
	unsigned	nFixturesRun;

	/*
	|| finally, the number of suites failed/run
	*/
	unsigned	nSuitesFailed;
	unsigned	nSuitesRun;
	};

/**** end of oversee_internals.h ****/
#endif /* OVERSEE_INTERNALS_H_ */


i/oversee_platform_types.h
/* oversee_platform_types.h	Thu Feb 28 2013 16:27:21 tmm */

/*

Module:  oversee_platform_types.h

Function:
	The "scalar types" for the oversee platform.

Version:
	V0.91a	Thu Feb 28 2013 16:27:21 tmm	Edit level 1

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY  14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore.
 
Author:
	Terry Moore	February 2013

Revision history:
   0.91a  Thu Feb 28 2013 16:27:21  tmm
	Module created.

*/

#ifndef OVERSEE_PLATFORM_TYPES_H__		/* prevent multiple includes */
#define OVERSEE_PLATFORM_TYPES_H__

#ifndef OVERSEE_TYPES_H__
# include "oversee_types.h"
#endif

/****************************************************************************\
|
|	Platform API functions
|
| Platform APIs are used by OVERSEE to do its work.  
|
\****************************************************************************/

typedef struct OVERSEE_HPLATFORM_s *OVERSEE_HPLATFORM;

typedef	
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_REFERENCE_FN)
void OVERSEE_ABI_STD
OVERSEE_PLATFORM_REFERENCE_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM	/* hPlatform */
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_DEREFERENCE_FN)
void OVERSEE_ABI_STD
OVERSEE_PLATFORM_DEREFERENCE_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM	/* hPlatform */
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_STRLEN_FN)
size_t OVERSEE_ABI_STD
OVERSEE_PLATFORM_STRLEN_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM	/* hPlatform */,
	OVERSEE_SAL_In_opt_z 
		const char *
	);

typedef	
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_SNPRINTF_FN)
int OVERSEE_ABI_VARARGS
OVERSEE_PLATFORM_SNPRINTF_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM		/* hPlatform */,
	OVERSEE_SAL_Out_writes_z(OVERSEE_SAL_Param(3))
	OVERSEE_SAL_Pre_defensive
		char * OVERSEE_RESTRICT		/* pBuffer */,
	OVERSEE_SAL_In
		size_t				/* nBuffer */,
	OVERSEE_SAL_In_z
	OVERSEE_SAL_Printf_format_string
	OVERSEE_SAL_Pre_defensive
		const char * OVERSEE_RESTRICT	/* pFormat */,
		...
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_STRCMP_FN)
int OVERSEE_ABI_STD
OVERSEE_PLATFORM_STRCMP_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM		/* hPlatform */,
	OVERSEE_SAL_In_z
	OVERSEE_SAL_Pre_defensive
		const char *			/* pS1 */,
	OVERSEE_SAL_In_z
	OVERSEE_SAL_Pre_defensive
		const char *			/* pS2 */
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_STRSTR_FN)
int OVERSEE_ABI_STD
OVERSEE_PLATFORM_STRSTR_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM		/* hPlatform */,
	OVERSEE_SAL_In_z
	OVERSEE_SAL_Pre_defensive
		const char *			/* pLookInThisString */,
	OVERSEE_SAL_In_z
	OVERSEE_SAL_Pre_defensive
		const char *			/* pStringToLookFor */
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_STRNCMP_FN)
int OVERSEE_ABI_STD
OVERSEE_PLATFORM_STRNCMP_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM		/* hPlatform */,
	OVERSEE_SAL_In_reads_or_z(OVERSEE_SAL_Param(4))
	OVERSEE_SAL_Pre_defensive
		const char *			/* pS1 */,
	OVERSEE_SAL_In_reads_or_z(OVERSEE_SAL_Param(4))
	OVERSEE_SAL_Pre_defensive
		const char *			/* pS2 */,
	OVERSEE_SAL_In
		size_t				/* n */
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_OPEN_RESULT_STREAM_FN)
void OVERSEE_ABI_STD
OVERSEE_PLATFORM_OPEN_RESULT_STREAM_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM		/* hPlatform  */,
	OVERSEE_SAL_Out_writes_bytes(OVERSEE_SAL_Param(3))
	OVERSEE_SAL_Pre_defensive
		OVERSEE_RESULT_STREAM_API *	/* pStreamAPIs */,
	OVERSEE_SAL_In
		size_t				/* sizeStreamAPIs */
	);

/* the function type for the parameter for protected call */
typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_PROTECTED_CALL_CB_FN)
void OVERSEE_ABI_STD
OVERSEE_PLATFORM_PROTECTED_CALL_CB_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM			/* hPlatform */,
	OVERSEE_SAL_In_opt
		void *					/* context */
	);

typedef 
OVERSEE_SAL_Function_class(OVERSEE_PLATFORM_PROTECTED_CALL_FN)
int OVERSEE_ABI_STD
OVERSEE_PLATFORM_PROTECTED_CALL_FN(
	OVERSEE_SAL_In
		OVERSEE_HPLATFORM			/* hPlatform */,
	OVERSEE_SAL_In OVERSEE_SAL_Pre_defensive
		OVERSEE_PLATFORM_PROTECTED_CALL_CB_FN *	/* pFN */,
	OVERSEE_SAL_In_opt
		void *					/* context */
	);

/**** end of oversee_platform_types.h ****/
#endif /* OVERSEE_PLATFORM_TYPES_H__ */


i/oversee.h
/* oversee.h	Mon Feb 11 2013 02:29:37 tmm */

/*

Module:  oversee.h

Function:
	Header file for embedded unit testing (loosely based on seatest).

Version:
	V0.91a	Mon Feb 11 2013 02:29:37 tmm	Edit level 1

Copyright notice:
	This file copyright (C) 2013 by

		Terrill Moore
		3170 Perry City Road
		Trumansburg, NY  14886

	An unpublished work.  All rights reserved.
	
	This file is proprietary information, and may not be disclosed or
	copied without the prior permission of Terrill Moore.
 
Author:
	Terry Moore	February 2013

Revision history:
   0.91a  Mon Feb 11 2013 02:29:37  tmm
	Module created.

*/

#ifndef OVERSEE_H__		/* prevent multiple includes */
#define OVERSEE_H__

#ifndef OVERSEE_TYPES_H__
# include "oversee_types.h"
#endif

/*

MODULE DOCUMENTATION:	Naming Conventions

All global-namespace symbols defined by OVERSEE begin with "OVERSEE".

Normal industry practice is to use a single leading underscore for symbols 
that are in the global namespace but are not API symbols. For exmaple,
header-file inclusion guards are often named _FOO_H_ (to prevent multiple
includes of foo.h).  However ISO C99 says that any symbol beginning with
_ and uppercase, or double-underscore, is reserved for the implementation.
So OVERSEE uses trailing underscores to distinguish such symbols. Header-
file guard symbols are named OVERSEE_{something}_H__.

For similar reasons, OVERSEE avoids the common practice of using a leading
underscore on a structure tag. Structure tags are all named 
OVERSEE_{STRUCTNAME}_s.  Similarly, union tags are OVERSEE_{UNIONNAME}_u, 
and enum tags are OVERSEE_{ENUM}_e.  So, for example, the declaration
of the OVERSEE_TEST structure is 

.	typedef struct OVERSEE_TEST_s OVERSEE_TEST;

rather than

.	typedef struct _OVERSEE_TEST OVERSEE_TEST;

When defining fields or variable names, we use a loose form of 
Hungarian notation. Names beginning with "p" are pointers. Names beginning
with "pv" are pointers to vectors. Names beginning with "u" are some kind
of unsigned type; names beginning with "i" are some kind of signed integral
type. Names beginning with "size" denote an unsigned size in bytes; names 
beginning with "length" or "n" denote an unsigned count of units.  Names
beginning with "f" are logically boolean flags.

Hungarian notation has been criticised on the basis that the type of the
variable should be declared and not repeated. This is true to an extent;
however, C is not really type agnostic. It's important to know whether a
variable is signed or unsigned when looking at code, and making the eye
wander to a declaration, especially of a structure in another file, is
both hard on the reader and distracting for purposes of analysis. We find
that readable software, like natural language, necessarily has a certain
amount of redundant information. Elegant writing uses this redundancy
consciously, for rhetorical effect. Coding style should manage redundancy
to guide the writer, the maintaininer, and the auditor.

*/



/****************************************************************************\
|
|	The various structures
|
\****************************************************************************/

struct OVESEE_TEST_s
	{
	OVERSEE_CONTEXT			*pContext;
	OVERSEE_FIXTURE			*pParentFixture;

	/*
	|| pTestId points to the name of the test. Normally this is taken
	|| from the test function name.
	*/
	const char			*pTestId;
	};

/*
|| A fixture represents the test fixture. In theory, there is one of these for
|| each actual fixture, but in fact these are all "virtual", and are only
|| intantiated for the execution life of the fixture.
*/
struct OVERSEE_FIXTURE_s
	{
	OVERSEE_CONTEXT			*pContext;
	OVERSEE_SUITE			*pParentSuite;
	OVERSEE_FIXTURE_SETUP_FN	*pSetupFn;
	void				*pSetupContext;
	OVERSEE_FIXTURE_TEARDOWN_FN	*pTeardownFn;
	void				*pTeardownContext;

	/*
	|| pFixtureId points to the name of the test fixture. Normally
	|| this is taken from the source file name.
	*/
	const char			*pFixtureId;
	};

struct OVERSEE_SUITE_s
	{
	OVERSEE_CONTEXT			*pContext;
	OVERSEE_FIXTURE			*pCurrentFixture;

	OVERSEE_SUITE_SETUP_FN		*pSetupFn;
	void				*pSetupContext;
	OVERSEE_SUITE_TEARDOWN_FN	*pTeardownFn;
	void				*pTeardownContext;	
	};

/****************************************************************************\
|
|	The API functions
|
\****************************************************************************/

OVERSEE_BEGIN_DECLS

OVERSEE_SAL_Ret_notnull
OVERSEE_SUITE *
Oversee_StaticInit(
	void
	);

OVERSEE_SAL_Success(return != 0)
OVERSEE_BOOL
OverseeSuite_RunTests(
	OVERSEE_SAL_In
		OVERSEE_SUITE *			/* pSuite */,
	OVERSEE_SAL_In
		OVERSEE_SUITE_DRIVER_FN *	/* pSuiteDriver */,
	OVERSEE_SAL_In_opt
		void *				/* pSuiteDriverCtx */
	);

OVERSEE_END_DECLS

/**** end of oversee.h ****/
#endif /* OVERSEE_H__ */
